cpp/inc/bond/ext/grpc/unary_call.h

```diff
@@ -87,163 +134,282 @@ namespace detail {
+class unary_call final : public detail::unary_call_base<TRequest, TResponse>
+{
+private:
+    using impl_type = detail::unary_call_impl<TRequest, TResponse>;
+    using base_type = detail::unary_call_base<TRequest, TResponse>;
 
-    /// @brief Responds to the client with the given message and status.
+public:
+    /// @brief The default constructor of unary_call is not callable: the
+    /// Bond-over-gRPC infrastructure will create unary_call instances as
+    /// needed.
     ///
-    /// Only the first call to \p Finish or \p FinishWithError will be
-    /// honored.
-    void Finish(const TResponse& msg, const grpc::Status& status = grpc::Status::OK)
-    {
-        Finish(bond::bonded<TResponse>{msg}, status);
-    }
+    /// Once created, a unary_call instance can only be moved to another
+    /// unary_call or converted to a \ref shared_unary_call (see the \ref
+    /// share function).
+    unary_call() = delete;
+
+    // unary_call is move-only
+    unary_call(const unary_call&) = delete;
+    unary_call& operator=(const unary_call&) = delete;
+
+    /// @brief Constructs a unary_call by moving from another unary_call.
+    ///
+    /// The moved-from instance is valid only for destruction and
+    /// assignment.
+    unary_call(unary_call&& other) = default;
 
-    /// @brief Responds to the client with the given message and status.
+    /// @brief Assign to this unary_call by moving from another unary_call.
     ///
-    /// Only the first call to \p Finish or \p FinishWithError will be
-    /// honored.
-    void Finish(const bond::bonded<TResponse>& msg, const grpc::Status& status = grpc::Status::OK)
+    /// The moved-from instance is valid only for destruction and
+    /// assignment.
+    unary_call& operator=(unary_call&& rhs) = default;
+
+    ~unary_call() = default;
+
+    void swap(unary_call& rhs) noexcept
     {
-        BOOST_ASSERT(_impl);
-        _impl->Finish(msg, status);
+        base_type::swap(rhs);
     }
 
-    /// @brief Responds to the client with the given status and no message.
+    /// @brief Creates a \ref shared_unary_call from this instance.
     ///
-    /// Only the first call to \p Finish or \p FinishWithError will be
-    /// honored.
-    void FinishWithError(const grpc::Status& status)
+    /// This instance will be in a moved-from state after \ref share has
+    /// been called, and the returned shared instance should be used to
+    /// finish the call.
+    shared_unary_call<TRequest, TResponse> share() &&
     {
-        BOOST_ASSERT(_impl);
-        _impl->FinishWithError(status);
+        return shared_unary_call<TRequest, TResponse>(std::move(*this));
     }
 
     template <typename SUCDRequest, typename SUCDResponse, typename SUCDThreadPool>
     friend struct detail::service_unary_call_data;
 
 private:
-    std::shared_ptr<detail::unary_call_impl<TRequest, TResponse>> _impl;
+    explicit unary_call(impl_type*&& impl) noexcept
+        : base_type(std::move(impl))
```

Comment:
Please remove the `&&` and do not `move`. Or even better, do `using base_type::base_type;` if we don't care about VC12 here.

Reply:
Thanks. That's a copy/paste mistake.